QEfficient/transformers/quantizers/quantizer_awq.py

```diff
@@ -0,0 +1,198 @@
+
+import copy
+
+import torch
+import torch.nn as nn
+from transformers.integrations.awq import AWQ_SCALES_MAPPINGS
+from transformers.quantizers.quantizer_awq import AwqQuantizer
+from transformers.utils.quantization_config import AwqBackendPackingMethod, AwqConfig, AWQLinearVersion
+
+from QEfficient.transformers.quantizers.awq import WQLinear_GEMM
+from QEfficient.utils.logging_utils import logger
+
+
+class QEffAwqConfig(AwqConfig):
+    def post_init(self):
+        """
+        Safety checker that arguments are correct
+        """
+
+        if self.backend not in [AwqBackendPackingMethod.AUTOAWQ]:
+            raise ValueError(
+                f"Only quantization backend {AwqBackendPackingMethod.AUTOAWQ} is supported - not recognized backend {self.backend}"
+            )
+
+        self.version = AWQLinearVersion.from_str(self.version)
+        if self.version not in [AWQLinearVersion.GEMM]:
+            raise ValueError(
+                f"Only {AWQLinearVersion.GEMM} version in supported - not recognized version {self.version}"
+            )
+
+        if self.do_fuse or self.fuse_max_seq_len is not None:
+            raise ValueError(
+                f"fused modules are not supported, got do_fuse={self.do_fuse}, fuse_max_seq_len={self.fuse_max_seq_len}"
+            )
+
+        if self.bits != 4:
+            raise ValueError(f"Only 4-bit AWQ quantization is supported, got bits={self.bits}")
+
+
+class QEffAwqQuantizer(AwqQuantizer):
+    def __init__(self, quantization_config: QEffAwqConfig, **kwargs):
+        super().__init__(quantization_config, **kwargs)
+
+    def validate_environment(self, device_map, **kwargs):
+        # No need to validate as we will always use pytorch CPU version.
+        return True
+
+    @property
+    def is_trainable(self):
+        return False
+
+    def update_torch_dtype(self, torch_dtype):
+        if torch_dtype not in [None, torch.float32]:
+            logger.warning(f"Requested dtype {torch_dtype} is not supported, overriding to None")
+        return None
+
+    def _process_model_before_weight_loading(self, model, **kwargs):
+        self.modules_to_not_convert = get_keys_to_not_convert(model)
+
+        if self.quantization_config.modules_to_not_convert is not None:
+            self.modules_to_not_convert.extend(self.quantization_config.modules_to_not_convert)
+
+        model, has_been_replaced = replace_linear_layer_with_awq_gemm(
+            model, quantization_config=self.quantization_config, modules_to_not_convert=self.modules_to_not_convert
+        )
+
+        model = replace_quantization_scales(model, model.config.model_type)
+        if not has_been_replaced:
+            logger.warning(
+                "You are loading an AWQ model but no linear modules were found in your model."
+                " Please double check your model architecture, or submit an issue on github if you think this is a bug."
+            )
+
+
+class ScaledActivation(nn.Module):
```

Comment:
If any of the functions or classes defined here can be reused from transformers, please do so. Ignore if already done.

Reply:
This class is defined in `auto-awq` package, We don't want to add extra package dependency for this.

Reply:
From [HF transformer docs](https://huggingface.co/docs/transformers/main/en/quantization/awq), it seems that autoawq is a prerequisite for running AWQ transformer models. I think we can add an optional dependency for the same.

Reply:
We don't want to add this because of legal review process. This will delay the awq feature release